wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/DevfileManager.java

```diff
@@ -121,23 +187,69 @@ public void resolveReference(DevfileImpl devfile, FileContentProvider fileConten
     }
   }
 
-  private DevfileImpl parse(String content, ValidationFunction validationFunction)
+  private DevfileImpl parse(
+      String content, ObjectMapper mapper, Map<String, String> overrideProperties)
       throws DevfileFormatException {
-    JsonNode parsed = validationFunction.validate(content);
-
     DevfileImpl devfile;
     try {
-      devfile = objectMapper.treeToValue(parsed, DevfileImpl.class);
+      JsonNode parsed = mapper.readTree(content);
+      parsed = applyPropertiesOverride(parsed, overrideProperties);
+      schemaValidator.validate(parsed);
+      devfile = mapper.treeToValue(parsed, DevfileImpl.class);
     } catch (JsonProcessingException e) {
       throw new DevfileFormatException(e.getMessage());
+    } catch (IOException e) {
+      throw new DevfileFormatException("Unable to parse Devfile. Error: " + e.getMessage());
     }
-
     integrityValidator.validateDevfile(devfile);
     return devfile;
   }
 
-  @FunctionalInterface
-  private interface ValidationFunction {
-    JsonNode validate(String content) throws DevfileFormatException;
+  private JsonNode applyPropertiesOverride(
+      JsonNode devfileNode, Map<String, String> overrideProperties) throws DevfileFormatException {
+    for (Map.Entry<String, String> entry : overrideProperties.entrySet()) {
+      // prepare stuff
+      String[] pathSegments = entry.getKey().split("\\.");
+      if (pathSegments.length < 1) {
+        continue;
+      }
+      String lastSegment = pathSegments[pathSegments.length - 1];
+      JsonNode currentNode = devfileNode;
+      // iterate until we reach last but one path segment
+      Iterator<String> pathSegmentsIterator =
+          asList(copyOf(pathSegments, pathSegments.length - 1)).iterator();
+      while (pathSegmentsIterator.hasNext()) {
+        String currentSegment = pathSegmentsIterator.next();
+        JsonNode result = currentNode.path(currentSegment);
+        if (result.isMissingNode()) {
+          currentNode = ((ObjectNode) currentNode).putObject(currentSegment);
```

Comment:
is this cast safe?

Reply:
In our case - yes since we're never get to the end value which can be TextNode or IntNode, we're always stay at least 1 level up.